### PR TITLE
[RDY] vim-patch:7.4.551

### PR DESCRIPTION
**Pull request by epaulson10**
```
Problem:    "ygn" may yank too much. (Fritzophrenic)  Issue 295.
Solution:   Check the width of the next match. (Christian Brabandt)
```

https://code.google.com/p/vim/source/detail?r=v7-4-551

Original patch:

``` diff
diff --git a/src/nvim/search.c b/src/nvim/search.c
--- a/src/nvim/search.c
+++ b/src/nvim/search.c
@@ -4441,12 +4441,11 @@ current_quote(oap, count, include, quote

 #endif /* FEAT_TEXTOBJ */

-static int is_one_char __ARGS((char_u *pattern));
+static int is_one_char __ARGS((char_u *pattern, int move));

 /*
  * Find next search match under cursor, cursor at end.
  * Used while an operator is pending, and in Visual mode.
- * TODO: redo only works when used in operator pending mode
  */
     int
 current_search(count, forward)
@@ -4491,7 +4490,7 @@ current_search(count, forward)
    orig_pos = pos = start_pos = curwin->w_cursor;

     /* Is the pattern is zero-width? */
-    one_char = is_one_char(spats[last_idx].pat);
+    one_char = is_one_char(spats[last_idx].pat, TRUE);
     if (one_char == -1)
     {
    p_ws = old_p_ws;
@@ -4550,6 +4549,10 @@ current_search(count, forward)
     start_pos = pos;
     flags = forward ? SEARCH_END : 0;

+    /* Check again from the current cursor position,
+     * since the next match might actually by only one char wide */
+    one_char = is_one_char(spats[last_idx].pat, FALSE);
+
     /* move to match, except for zero-width matches, in which case, we are
      * already on the next match */
     if (!one_char)
@@ -4599,26 +4602,38 @@ current_search(count, forward)

 /*
  * Check if the pattern is one character or zero-width.
+ * If move is TRUE, check from the beginning of the buffer, else from the
+ * current cursor position.
  * Returns TRUE, FALSE or -1 for failure.
  */
     static int
-is_one_char(pattern)
+is_one_char(pattern, move)
     char_u *pattern;
+    int        move;
 {
     regmmatch_T    regmatch;
     int        nmatched = 0;
     int        result = -1;
     pos_T  pos;
     int        save_called_emsg = called_emsg;
+    int        flag = 0;

     if (search_regcomp(pattern, RE_SEARCH, RE_SEARCH,
                          SEARCH_KEEP, &regmatch) == FAIL)
    return -1;

     /* move to match */
-    clearpos(&pos);
+    if (move)
+   clearpos(&pos)
+    else
+    {
+   pos = curwin->w_cursor;
+   /* accept a match at the cursor position */
+   flag = SEARCH_START;
+    }
+
     if (searchit(curwin, curbuf, &pos, FORWARD, spats[last_idx].pat, 1,
-                    SEARCH_KEEP, RE_SEARCH, 0, NULL) != FAIL)
+                 SEARCH_KEEP + flag, RE_SEARCH, 0, NULL) != FAIL)
     {
    /* Zero-width pattern should match somewhere, then we can check if
     * start and end are in the same position. */
diff --git a/src/nvim/testdir/test53.in b/src/nvim/testdir/test53.in
--- a/src/nvim/testdir/test53.in
+++ b/src/nvim/testdir/test53.in
@@ -79,6 +79,8 @@ ggdgn.
 :" test repeating gUgn
 /^Depp
 gggUgn.
+gg/a:0\@!\zs\d\+
+nygnop
 :/^start:/,/^end:/wq! test.out
 ENDTEST

@@ -108,6 +110,11 @@ delete first and last chars
 uniquepattern uniquepattern
 my very excellent mother just served us nachos
 for (i=0; i<=10; i++)
+a:10
+
+a:1
+
+a:20
 Y
 text
 Y
diff --git a/src/nvim/testdir/test53.ok b/src/nvim/testdir/test53.ok
--- a/src/nvim/testdir/test53.ok
+++ b/src/nvim/testdir/test53.ok
@@ -49,6 +49,12 @@ elete first and last char
  uniquepattern
 my very excellent mongoose just served us nachos
 for (j=0; i<=10; i++)
+a:10
+
+a:1
+1
+
+a:20

 text
 Y
diff --git a/src/nvim/version.c b/src/nvim/version.c
--- a/src/nvim/version.c
+++ b/src/nvim/version.c
@@ -742,6 +742,8 @@ static char *(features[]) =
 static int included_patches[] =
 {   /* Add new patch number below this line */
 /**/
+    551,
+/**/
     550,
 /**/
     549,
```

### CODE REVIEW

File: src/nvim/search.c [hunk 1]
```diff
@@ -3878,24 +3881,34 @@ current_search (
 
 /*
  * Check if the pattern is one character or zero-width.
+ * If move is TRUE, check from the beginning of the buffer,
+ * else from the current cursor position.
  * Returns TRUE, FALSE or -1 for failure.
  */
-static int is_one_char(char_u *pattern)
+static int is_one_char(char_u *pattern, int move)
```
Comment (fwalch) [inline src/nvim/search.c]: You could change this to `bool` and use `true`/`false` instead of `TRUE`/`FALSE` in the invocations and the added comments.

**Comment by fwalch** [comment]
One comment about `bool`, otherwise looks good to me! :+1:

Would you mind signing the [CLA](https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#pull-requests)?

**Comment by epaulson10** [comment]
Thanks for the feedback, I went ahead and changed those to bool and I signed the CLA. Cheers!

File: src/nvim/search.c [hunk 2]
```diff
@@ -3878,24 +3881,34 @@ current_search (
...
   if (search_regcomp(pattern, RE_SEARCH, RE_SEARCH,
           SEARCH_KEEP, &regmatch) == FAIL)
     return -1;
 
   /* move to match */
-  clearpos(&pos);
+  if (move) {
+      clearpos(&pos);
+  }
+  else {
```
Comment (fwalch) [inline src/nvim/search.c]: Sorry to be so nit-picky, but I just noticed that the indentation seems to be 4 spaces. It should be two spaces according to the [style guide](http://neovim.org/develop/style-guide.xml). Also, while I can't actually find it in the style guide right now, you can also move up the else (like `} else {`). This is what's used [elsewhere](https://github.com/neovim/neovim/commit/2e873c8767056e75e8e44d878caf5673df06f423#diff-a6109b430c5b2379023dc39efac014dcR330) in the codebase.

**Comment by fwalch** [comment]
Thanks! Added one minor comment about the code style. Can you [squash the commits into one and force-push](https://help.github.com/articles/about-git-rebase/)? Then I think this should be ready to get merged.

**Comment by epaulson10** [comment]
I fixed the indentation and squashed the commits together.

**Comment by justinmk** [comment]
Thanks

**Pull request merged**